base_layer/core/src/chain_storage/postgres_db/models/tx_outputs.rs

```diff
@@ -0,0 +1,165 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    blocks,
+    blocks::BlockHash,
+    chain_storage::{
+        postgres_db::{models::error::PostgresError, schema::*},
+        DbKeyValuePair,
+        DbValue,
+    },
+    transactions::types::{BlindingFactor, HashOutput},
+};
+use chrono::{NaiveDateTime, Utc};
+use diesel::{self, expression::dsl, prelude::*, OptionalExtension};
+use log::*;
+use serde_json::Value;
+use std::convert::{TryFrom, TryInto};
+use tari_crypto::tari_utilities::{hex::Hex, Hashable};
+
+const LOG_TARGET: &str = "b::c::storage::postgres:tx_outputs";
+
+#[derive(Queryable, Identifiable, Insertable)]
+#[table_name = "tx_outputs"]
+#[primary_key(hash)]
+pub struct TxOutput {
+    hash: String,
+    features_flags: i16,
+    features_maturity: i64,
+    commitment: String,
+    proof: Vec<u8>,
+    tx_output: String,
+    spent: String,
+}
+
+impl TxOutput {
+    pub fn insert_if_not_exists(output: &TransactionOutput, conn: &PgConnection) -> Result<bool, PostgresError> {
+        let hash = output.hash();
+
+        if TxOutput::fetch(&hash, conn)?.is_some() {
+            warn!(
+                target: LOG_TARGET,
+                "Tried to insert transactional output with hash:{} but it already exists",
+                hash.to_hex()
+            );
+
+            return Ok(false);
+        }
+
+        let row: TxOutput = output.try_into()?;
+        if row.hash != hash.to_hex() {
+            return Err(PostgresError::Other("tx and tx hash don't match".to_string()));
+        }
+
+        diesel::insert_into(tx_outputs::table)
+            .values(&row)
+            .execute(conn)
+            .map_err(|e| PostgresError::CouldNotAdd(e.to_string()))?;
+
+        Ok(true)
+    }
+
+    pub fn fetch(hash: &HashOutput, conn: &PgConnection) -> Result<Option<TxOutput>, PostgresError> {
+        let mut results: Vec<TxOutput> = tx_outputs::table
+            .filter(tx_outputs::hash.eq(hash.to_hex()))
+            .get_results(conn)
+            .map_err(|e| PostgresError::NotFound(e.to_string()))?;
+
+        Ok(results.pop())
+    }
+
+    pub fn fetch_output(hash: &HashOutput, conn: &PgConnection) -> Result<Vec<TxOutput>, PostgresError> {
```

Comment:
Please doc all these functions. They're almost exactly the same, so let the user / client know what they do.
This one might be better named `fetch_block_inputs`

Reply:
changed